File: leetcode/687.longest-univalue-path.py
# Definition for a binary tree node.
# class TreeNode(object):
#     def __init__(self, x):
#         self.val = x
#         self.left = None
#         self.right = None

class Solution(object):
    def __init__(self):
        self.res = 0
    
    def longestUnivaluePath(self, root):
        """
        :type root: TreeNode
        :rtype: int
        """
        self.helper(root)
        return self.res
        
    def helper(self, cur):
        if cur is None:
            return 0
        
        left = self.helper(cur.left)
        right = self.helper(cur.right)
        
        left = left + 1 if cur.left is not None and cur.left.val == cur.val else 0
        right = right + 1 if cur.right is not None and cur.right.val == cur.val else 0
            
        self.res = max(self.res, left + right)
            
        return max(left, right)
        

File: leetcode/687.longest-univalue-path.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
class Solution {
private:
    int ans = 0;
    
    int helper(TreeNode* cur)
    {
        if (cur == NULL)
        {
            return 0;
        }
        
        int left = helper(cur->left);
        int right = helper(cur->right);
        
        left = (cur->left && cur->left->val == cur->val) ? left + 1 : 0;
        right = (cur->right && cur->right->val == cur->val) ? right + 1 : 0;
        
        ans = max(ans, left + right);
        
        return max(left, right);
    }
    
public:
    int longestUnivaluePath(TreeNode* root) {
        helper(root);
        return ans;
    }
};